File: unittests/backends/quake_backend/test_loop_app.cpp
/*******************************************************************************
 * Copyright (c) 2026 NVIDIA Corporation & Affiliates.                         *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/

#include <cudaq.h>

#include <cmath>
#include <iostream>
#include <vector>

// IMPORTANT: loop-count annotations are part of this test contract. Kernel
// names ending in `_expected_N_loop(s)` tell the quake fake mock server how
// many `cc.loop` operations must be present in the submitted client payload
// before the server does its own execution-time full unroll.

// Expected remaining payload loops: none. Broadcast X over the whole qvector
// must lower to static wire operations in the submitted payload.
__qpu__ int broadcast_x_qvector_expected_0_loops() {
  cudaq::qvector q(3);
  x(q);

  int result = 0;
  auto bits = mz(q);
  if (bits[0])
    result |= 1;
  if (bits[1])
    result |= 2;
  if (bits[2])
    result |= 4;
  return result;
}

// Expected remaining payload loops: none. The whole-register control must be
// expanded before wire assignment so all control wires are static.
__qpu__ int whole_register_control_x_expected_0_loops() {
  cudaq::qvector ctrls(3);
  cudaq::qubit target;

  x(ctrls);
  x<cudaq::ctrl>(ctrls, target);
  return mz(target);
}

// Expected remaining payload loop: the outer pass loop. The loops that index
// quantum data (`q[i]`) are unrolled so wire IDs are static in the payload.
__qpu__ int loop_payload_stress_expected_1_loop() {
  constexpr int width = 4;
  cudaq::qvector q(width);

  // KEEP: this classical repeat loop does not index quantum data directly.
  for (int pass = 0; pass < 2; ++pass)
    // UNROLL: this loop indexes `q[i]`, so wire IDs must become static.
    for (int i = 0; i < width; ++i)
      x(q[i]);

  int i = 0;
  // UNROLL: this loop indexes `q[i]`, so wire IDs must become static.
  while (i < width) {
    x(q[i]);
    ++i;
  }

  int result = 0;
  if (mz(q[0]))
    result |= 1;
  if (mz(q[1]))
    result |= 2;
  if (mz(q[2]))
    result |= 4;
  if (mz(q[3]))
    result |= 8;

  return result;
}

// Expected remaining payload loop: the loop over measurement results. This loop
// does not access quantum data, so it should stay rolled.
__qpu__ int measurement_result_loop_expected_1_loop() {
  constexpr int width = 4;
  cudaq::qvector q(width);

  x(q[0]);
  x(q[2]);

  auto bits = mz(q);
  int result = 0;
  // KEEP: this loop only indexes measurement results, not quantum data.
  for (int i = 0; i < width; ++i)
    if (bits[i])
      result |= 1 << i;

  return result;
}

// Expected remaining payload loops: outer, group, classical accumulation, and
// measurement-result accumulation. The quantum-data loop over `q[wire]` is
// unrolled for static wire IDs.
__qpu__ int nested_mixed_loop_payload_expected_4_loops() {
  constexpr int width = 4;
  cudaq::qvector q(width);

  int classical = 0;
  // KEEP: this classical outer loop does not index quantum data directly.
  for (int outer = 0; outer < 1; ++outer)
    // KEEP: this classical group loop does not index quantum data directly.
    for (int group = 0; group < 2; ++group) {
      // UNROLL: this loop indexes `q[wire]`, so wire IDs must become static.
      for (int wire = 0; wire < width; ++wire)
        x(q[wire]);

      // KEEP: this loop is classical accumulation only.
      for (int i = 0; i < 3; ++i)
        classical += group + i;
    }

  x(q[0]);

  auto bits = mz(q);
  int result = classical;
  // KEEP: this loop only iterates over measurement results.
  for (auto bit : bits)
    if (bit)
      result += 16;

  return result;
}

// Expected remaining payload loop: the phase-kickback repeat loop. This is the
// fixed-wire core of phase estimation: repeated controlled unitary powers on a
// prepared eigenstate. The loop should stay rolled because it uses fixed
// qubits.
__qpu__ int fixed_wire_phase_estimation_expected_1_loop() {
  cudaq::qubit counting, eigenstate;

  x(eigenstate);
  h(counting);

  // KEEP: repeated controlled rotations over fixed qubits should stay rolled.
  for (int power = 0; power < 3; ++power)
    r1<cudaq::ctrl>(M_PI, counting, eigenstate);

  h(counting);
  return mz(counting);
}

// Expected remaining payload loops: none. The triangular inner bound depends on
// the outer induction variable, and both loops index quantum data. Both loops
// must unroll so the client payload contains only static wire operations.
__qpu__ int parent_dependent_triangular_expected_0_loops() {
  constexpr int width = 4;
  cudaq::qvector q(width);

  x(q[0]);

  // UNROLL: `i` indexes quantum data and bounds the inner loop.
  for (int i = 0; i < width - 1; ++i)
    // UNROLL: `j` indexes quantum data with a parent-dependent lower bound.
    for (int j = i + 1; j < width; ++j)
      x<cudaq::ctrl>(q[i], q[j]);

  int result = 0;
  if (mz(q[0]))
    result |= 1;
  if (mz(q[1]))
    result |= 2;
  if (mz(q[2]))
    result |= 4;
  if (mz(q[3]))
    result |= 8;

  return result;
}

// Expected remaining payload loop: the outer repeat loop. The inner triangular
// parent-dependent quantum loops must unroll, but the outer repeat does not
// index quantum data directly and should stay rolled.
__qpu__ int repeated_parent_dependent_triangular_expected_1_loop() {
  constexpr int width = 4;
  cudaq::qvector q(width);

  x(q[0]);

  // KEEP: this repeat loop does not index quantum data directly.
  for (int pass = 0; pass < 2; ++pass)
    // UNROLL: `i` indexes quantum data and bounds the inner loop.
    for (int i = 0; i < width - 1; ++i)
      // UNROLL: `j` indexes quantum data with a parent-dependent lower bound.
      for (int j = i + 1; j < width; ++j)
        x<cudaq::ctrl>(q[i], q[j]);

  int result = 0;
  if (mz(q[0]))
    result |= 1;
  if (mz(q[1]))
    result |= 2;
  if (mz(q[2]))
    result |= 4;
  if (mz(q[3]))
    result |= 8;

  return result;
}

// Expected remaining payload loops: none. The middle loop does not directly
// index quantum data, but the inner loop indexes q[outer]. Unrolling must
// propagate through the non-blocking middle loop so every quantum access has a
// static wire ID in the submitted payload.
__qpu__ int transitive_parent_index_access_expected_0_loops() {
  constexpr int width = 3;
  cudaq::qvector q(width);

  // UNROLL: `outer` is used by the innermost quantum access q[outer].
  for (int outer = 0; outer < width; ++outer)
    // UNROLL: this loop is not directly blocking, but it encloses the
    // parent-dependent quantum access exposed by unrolling the inner loop.
    for (int middle = 0; middle < 3; ++middle)
      // UNROLL: this loop contains the q[outer] access that initially blocks
      // wire conversion.
      for (int inner = 0; inner < 3; ++inner)
        x(q[outer]);

  int result = 0;
  if (mz(q[0]))
    result |= 1;
  if (mz(q[1]))
    result |= 2;
  if (mz(q[2]))
    result |= 4;

  return result;
}

static bool checkResults(const std::vector<int> &results, int expected,
                         const char *name) {
  for (auto result : results) {
    if (result != expected) {
      std::cerr << name << " expected " << expected << ", got " << result
                << "\n";
      return false;
    }
  }
  return true;
}

int main() {
  if (!checkResults(cudaq::run(3, broadcast_x_qvector_expected_0_loops), 7,
                    "broadcast_x_qvector_expected_0_loops"))
    return 1;

  if (!checkResults(cudaq::run(3, whole_register_control_x_expected_0_loops), 1,
                    "whole_register_control_x_expected_0_loops"))
    return 1;

  if (!checkResults(cudaq::run(3, loop_payload_stress_expected_1_loop), 15,
                    "loop_payload_stress_expected_1_loop"))
    return 1;

  if (!checkResults(cudaq::run(3, measurement_result_loop_expected_1_loop), 5,
                    "measurement_result_loop_expected_1_loop"))
    return 1;

  if (!checkResults(cudaq::run(3, nested_mixed_loop_payload_expected_4_loops),
                    25, "nested_mixed_loop_payload_expected_4_loops"))
    return 1;

  if (!checkResults(cudaq::run(3, fixed_wire_phase_estimation_expected_1_loop),
                    1, "fixed_wire_phase_estimation_expected_1_loop"))
    return 1;

  if (!checkResults(cudaq::run(3, parent_dependent_triangular_expected_0_loops),
                    3, "parent_dependent_triangular_expected_0_loops"))
    return 1;

  if (!checkResults(
          cudaq::run(3, repeated_parent_dependent_triangular_expected_1_loop),
          5, "repeated_parent_dependent_triangular_expected_1_loop"))
    return 1;

  if (!checkResults(
          cudaq::run(3, transitive_parent_index_access_expected_0_loops), 7,
          "transitive_parent_index_access_expected_0_loops"))
    return 1;

  std::cout << "Loop payload stress passed.\n";
  return 0;
}

File: unittests/nvqpp/backends/quake_backend/test_loop_app.py
# ============================================================================ #
# Copyright (c) 2022 - 2026 NVIDIA Corporation & Affiliates.                   #
# All rights reserved.                                                         #
#                                                                              #
# This source code and the accompanying materials are made available under     #
# the terms of the Apache License 2.0 which accompanies this distribution.     #
# ============================================================================ #
import cudaq
import sys

cudaq.set_target("quake_fake")

width = 4

# IMPORTANT: loop-count annotations are part of this test contract. Kernel
# names ending in `_expected_N_loop(s)` tell the quake fake mock server how
# many `cc.loop` operations must be present in the submitted client payload
# before the server does its own execution-time full unroll.


# Expected remaining payload loops: the repeat loop, plus the generated
# measurement-result reduction from `to_integer(to_bools(mz(q)))`. The loops
# that index quantum data (`q[i]`) are unrolled so wire IDs are static.
@cudaq.kernel
def loop_payload_stress_expected_2_loops() -> int:
    q = cudaq.qvector(width)

    # KEEP: this classical repeat loop does not index quantum data directly.
    for repeat in range(2):
        # UNROLL: this loop indexes `q[i]`, so wire IDs must become static.
        for i in range(width):
            x(q[i])

    i = 0
    # UNROLL: this loop indexes `q[i]`, so wire IDs must become static.
    while i < width:
        x(q[i])
        i += 1

    # KEEP: this expression generates a measurement-result reduction loop.
    return cudaq.to_integer(cudaq.to_bools(mz(q)))


# Expected remaining payload loop: the loop over measurement results. This loop
# does not access quantum data, so it should stay rolled.
@cudaq.kernel
def measurement_result_loop_expected_1_loop() -> int:
    q = cudaq.qvector(width)

    x(q[0])
    x(q[2])

    bits = mz(q)
    result = 0
    # KEEP: this loop only indexes measurement results, not quantum data.
    for i in range(width):
        if bits[i]:
            result |= 1 << i

    return result


# Expected remaining payload loops: outer, group, classical accumulation, and
# measurement-result accumulation. The quantum-data loop over `q[wire]` is
# unrolled for static wire IDs.
@cudaq.kernel
def nested_mixed_loop_payload_expected_4_loops() -> int:
    q = cudaq.qvector(width)

    classical = 0
    # KEEP: this classical outer loop does not index quantum data directly.
    for outer in range(1):
        # KEEP: this classical group loop does not index quantum data directly.
        for group in range(2):
            # UNROLL: this loop indexes `q[wire]`, so wire IDs become static.
            for wire in range(width):
                x(q[wire])

            # KEEP: this loop is classical accumulation only.
            for i in range(3):
                classical += group + i

    x(q[0])

    bits = mz(q)
    result = classical
    # KEEP: this loop only indexes measurement results, not quantum data.
    for i in range(width):
        if bits[i]:
            result += 16

    return result


# Expected no remaining payload loops. The inner quantum-indexing loop has a
# parent-dependent trip count, so the outer loop must unroll first.
@cudaq.kernel
def parent_dependent_qubit_indexing_expected_0_loops() -> int:
    q = cudaq.qvector(width)

    # UNROLL: this loop resolves the inner trip count and base qubit index.
    for group in range(2):
        # UNROLL: this loop indexes `q[group + wire]`.
        for wire in range(width - group):
            x(q[group + wire])

    return 13


# Expected no remaining payload loops. The inner Pauli-indexing loop has a
# parent-dependent trip count, so the outer loop must unroll first.
@cudaq.kernel
def parent_dependent_pauli_indexing_expected_0_loops(
    angles: list[float],
    paulis: list[cudaq.pauli_word],
) -> int:
    q = cudaq.qvector(width)

    # UNROLL: this loop resolves the inner trip count and base Pauli index.
    for group in range(2):
        # UNROLL: this loop selects `paulis[group + offset]`.
        for offset in range(len(paulis) - group):
            term = group + offset
            exp_pauli(angles[term], q, paulis[term])

    return 17


# Expected remaining payload loop: the outer Trotter repetition. The orbital
# and Pauli-word selection loops must unroll before target decomposition.
@cudaq.kernel
def trotter_expected_1_loop(
    num_qubits: int,
    electron_count: int,
    angles: list[float],
    paulis: list[cudaq.pauli_word],
    trotter_steps: int,
) -> int:
    q = cudaq.qvector(num_qubits)
    # UNROLL: this loop indexes `q[orbital]`, so wire IDs must become static.
    for orbital in range(electron_count):
        x(q[orbital])

    # KEEP: this outer loop only repeats operations on statically resolved wires.
    for _ in range(trotter_steps):
        # UNROLL: this loop selects Pauli words using its induction variable.
        for term in range(len(angles)):
            exp_pauli(angles[term], q, paulis[term])

    return 7


# Expected remaining payload loop: the outer repetition. The nested composed
# index and reverse-index `while` loops must unroll to resolve each Pauli word.
@cudaq.kernel
def complex_pauli_indexing_expected_1_loop(
    angles: list[float],
    paulis: list[cudaq.pauli_word],
) -> int:
    q = cudaq.qvector(width)

    # KEEP: this outer loop only repeats statically resolvable Pauli operations.
    for _ in range(2):
        # UNROLL: this loop contributes to the nested Pauli-word index.
        for group in range(2):
            # UNROLL: this loop completes the nested Pauli-word index.
            for offset in range(2):
                term = 2 * group + offset
                exp_pauli(angles[term], q, paulis[term])

    term = 0
    # UNROLL: this loop selects Pauli words through a computed reverse index.
    while term < len(paulis):
        reverse_term = len(paulis) - term - 1
        exp_pauli(angles[reverse_term], q, paulis[reverse_term])
        term += 1

    return 11


def check_results(kernel, expected, *args):
    results = cudaq.run(kernel, *args, shots_count=3)
    assert len(results) == 3
    for result in results:
        assert result == expected, f"expected {expected}, got {result}"


try:
    check_results(loop_payload_stress_expected_2_loops, 15)
    check_results(measurement_result_loop_expected_1_loop, 5)
    check_results(nested_mixed_loop_payload_expected_4_loops, 25)
    check_results(parent_dependent_qubit_indexing_expected_0_loops, 13)
    check_results(parent_dependent_pauli_indexing_expected_0_loops, 17,
                  [0.1, 0.2, 0.3, 0.4], [
                      cudaq.pauli_word("ZZII"),
                      cudaq.pauli_word("IXXI"),
                      cudaq.pauli_word("YIYI"),
                      cudaq.pauli_word("IIIZ")
                  ])
    check_results(trotter_expected_1_loop, 7, 4, 2, [0.1, 0.2],
                  [cudaq.pauli_word("ZZII"),
                   cudaq.pauli_word("IXXI")], 2)
    check_results(complex_pauli_indexing_expected_1_loop, 11,
                  [0.1, 0.2, 0.3, 0.4], [
                      cudaq.pauli_word("ZZII"),
                      cudaq.pauli_word("IXXI"),
                      cudaq.pauli_word("YIYI"),
                      cudaq.pauli_word("IIIZ")
                  ])
except Exception as e:
    print(e)
    sys.exit(1)
